### PR TITLE
MPCA pipeline

### PR DESCRIPTION
**Pull request by sz144**
PR for [MPCA pipeline Card](https://github.com/pykale/pykale/projects/9#card-56270742).

### Description
Pipeline MPCA->Feature selection by Fisher score->SVM/Logistic Regression .

### Status
**Ready**

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [x] New tests added to cover the changes.
- [x] In-line docstrings updated and documentation `docs` updated.

### CODE REVIEW

File: kale/pipeline/mpca_trainer.py [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+"""
+
+import logging
+
+import numpy as np
+from sklearn.base import BaseEstimator, ClassifierMixin
+from sklearn.feature_selection import f_classif
+from sklearn.linear_model import LogisticRegression
+from sklearn.model_selection import GridSearchCV
+from sklearn.svm import SVC
```
Comment (haipinglu) [inline kale/pipeline/mpca_trainer.py:16]: Have you considered [`LinearSVC` (liblinear)](https://scikit-learn.org/stable/modules/generated/sklearn.svm.LinearSVC.html)? Or, can we make it an option (with one as the default)? SVC is using libsvm but LinearSVC is using liblinear.

I noticed this when checking https://scikit-learn.org/stable/modules/feature_selection.html and found that the examples using `LinearSVC` rather than `SVC`.

Did you check the matlab SVM I sent to you to see which version was used?

File: kale/pipeline/mpca_trainer.py [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+from sklearn.utils.validation import check_is_fitted
+
+from ..embed.mpca import MPCA
+
+classifiers = {"svc": [SVC, {"kernel": ["linear"], "C": np.logspace(-3, 2, 6)}],
```
Comment (haipinglu) [inline kale/pipeline/mpca_trainer.py:21]: Have you checked the Matlab code options? 
Does this cover the options used in our CMR Matlab code? It seems that we used the default there, which seems to be 1/n (an adaptive value) according to https://uk.mathworks.com/help/stats/fitclinear.html#d123e312449
At least this value of `C` should be covered.

File: kale/pipeline/mpca_trainer.py [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+               "lr": [LogisticRegression, {"C": np.logspace(-3, 2, 6)}]}
```
Comment (haipinglu) [inline kale/pipeline/mpca_trainer.py:22]: Again, consider to define repeated values `np.logspace(-3, 2, 6)` as a (global) variable

File: kale/pipeline/mpca_trainer.py [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+
+default_search_params = {'cv': 5}
```
Comment (haipinglu) [inline kale/pipeline/mpca_trainer.py:24]: We did not do cv in matlab, see https://uk.mathworks.com/help/stats/fitclinear.html#d123e314573
Is it an option here or a must? It may not be necessary, considering how Matlab deals with it (and we get better results). CV on small sample may overfit as well.

File: kale/pipeline/mpca_trainer.py [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+
+    def __init__(self, classifier='svc', classifier_params='auto', mpca_params=None,
+                 n_features=None, search_params=None):
+        """Trainer of pipeline: MPCA->Feature selection->Classifier
+
+        Args:
+            classifier (str, optional): Classifier for training. Options: support vector machine (svc) or
+                logistic regression (lr). Defaults to 'svc'.
+            classifier_params (dict, optional): Parameters of classifier. Defaults to 'auto'.
+            mpca_params (dict, optional): Parameters of Multi-linear PCA. Defaults to None.
```
Comment (haipinglu) [inline kale/pipeline/mpca_trainer.py:38]: Why Multi-linear here? Be consistent.

File: kale/pipeline/mpca_trainer.py [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+            self.search_params = default_search_params
+        else:
+            self.search_params = search_params
+
+        self.auto_classifier_param = False
+        if classifier_params == "auto":
+            self.auto_classifier_param = True
+            clf_param_gird = classifiers[classifier][1]
+            self.grid_search = GridSearchCV(classifiers[classifier][0](),
+                                            param_grid=clf_param_gird,
```
Comment (haipinglu) [inline kale/pipeline/mpca_trainer.py:69]: typo: gird

File: kale/pipeline/mpca_trainer.py [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+        x_proj = self.mpca.transform(x)
+
+        # feature selection
+        if self.n_features is None:
+            self.n_features = x_proj.shape[0]
+            self.feature_order = self.mpca.idx_order
+        else:
+            f_score, p_val = f_classif(x_proj, y)
+            self.feature_order = (-1 * f_score).argsort()
+        x_proj = x_proj[:, self.feature_order][:, :self.n_features]
```
Comment (haipinglu) [inline kale/pipeline/mpca_trainer.py:101]: Will a new name be better for selected features from x_proj?

File: kale/pipeline/mpca_trainer.py [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+
+        Args:
+            x (array-like tensor): input data, shape (n_samples, I_1, I_2, ..., I_N)
+
+        Returns:
+            array-like: decision scores, shape (n_samples, n_class)
+        """
+        return self.clf.decision_function(self.feature_extract(x))
+
+    def predict_proba(self, x):
```
Comment (haipinglu) [inline kale/pipeline/mpca_trainer.py:134]: Is `proba` more commonly used than `prob` for probability?

File: kale/pipeline/mpca_trainer.py [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+        Args:
+            x (array-like tensor): input data, shape (n_samples, I_1, I_2, ..., I_N)
+
+        Returns:
+            array-like: n_new, shape (n_samples, n_features)
+        """
+        check_is_fitted(self.clf)
+
+        x_proj = self.mpca.transform(x)
+        x_new = x_proj[:, self.feature_order][:, :self.n_features]
```
Comment (haipinglu) [inline kale/pipeline/mpca_trainer.py:157]: Use consistent naming convention, here x_new but you reused x_proj above.

File: tests/pipeline/test_mpca_trainer.py [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+
+
+@pytest.mark.parametrize("classifier", CLASSIFIERS)
+@pytest.mark.parametrize("params", PARAMS)
+def test_mpca_trainer(classifier, params):
+    trainer = MPCATrainer(classifier=classifier, **params)
+    trainer.fit(x, y)
+    y_pred = trainer.predict(x)
+    testing.assert_equal(np.unique(y), np.unique(y_pred))
+    assert accuracy_score(y, y_pred) >= 0.8
```
Comment (haipinglu) [inline tests/pipeline/test_mpca_trainer.py:27]: Expected training error (to be >0.8)?

File: tests/pipeline/test_mpca_trainer.py [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+
+    y_proba = trainer.predict_proba(x)
```
Comment (haipinglu) [inline tests/pipeline/test_mpca_trainer.py:29]: Does sklearn use `proba` for probalibity?

File: kale/pipeline/mpca_trainer.py (hunk 1 above)
Comment (sz144) [reply kale/pipeline/mpca_trainer.py:16]: SVC object has predict_proba function, which can give the probbabilitiy of each class, while LinearSVC does not have it.

File: kale/pipeline/mpca_trainer.py (hunk 4 above)
Comment (sz144) [reply kale/pipeline/mpca_trainer.py:24]: CV is used to determine the value of `C` only if `classifier_param` is set to be "auto".

File: kale/pipeline/mpca_trainer.py (hunk 7 above)
Comment (sz144) [reply kale/pipeline/mpca_trainer.py:101]: Is x_train better?

File: kale/pipeline/mpca_trainer.py (hunk 2 above)
Comment (sz144) [reply kale/pipeline/mpca_trainer.py:21]: There is a close value in `np.logspace(-3, 2, 6)`, which is a list `[0.001, 0.01, 0.1, 0, 1, 10]`. The optimal value of `C` will be determined by grid search if `classifier_params="auto"`. I will add 1/n to this list if necessary.

File: kale/pipeline/mpca_trainer.py (hunk 5 above)
Comment (sz144) [reply kale/pipeline/mpca_trainer.py:38]: Will change

File: tests/pipeline/test_mpca_trainer.py (hunk 10 above)
Comment (sz144) [reply tests/pipeline/test_mpca_trainer.py:27]: Training accuracy > 0.8, i.e. training error < 0.2.

File: kale/pipeline/mpca_trainer.py (hunk 1 above)
Comment (haipinglu) [reply kale/pipeline/mpca_trainer.py:16]: **If** LinearSVC can give a better accuracy and the user does not care the probability, then it is still a decent choice. How often did we report the probability of each class in our papers so far?

I understand probability is something good, but I do not consider it as essential or must-have.

File: kale/pipeline/mpca_trainer.py (hunk 4 above)
Comment (haipinglu) [reply kale/pipeline/mpca_trainer.py:24]: OK. Some light documentation/comments in the `tests` will be helpful for review / reading / future changes.

File: kale/pipeline/mpca_trainer.py (hunk 7 above)
Comment (haipinglu) [reply kale/pipeline/mpca_trainer.py:101]: Who can tell the difference? Isn't x also x_train?

File: kale/pipeline/mpca_trainer.py (hunk 2 above)
Comment (haipinglu) [reply kale/pipeline/mpca_trainer.py:21]: @sz144 1/n is an interesting and smart choice because when n --> infinity, 1/n --> 0 and no regularisation is needed.

**Review by sz144 — COMMENTED** [review]
Thanks for the comments.

File: kale/pipeline/mpca_trainer.py (hunk 1 above)
Comment (sz144) [reply kale/pipeline/mpca_trainer.py:16]: Probability is a feature requested by Cameron. How about making it optional (`LinearSVC` and `SVC(kernel="linear")`)?

Comment (haipinglu) [reply kale/pipeline/mpca_trainer.py]: A feature requested by a user does not mean to enforce it for all users. Otherwise, sklearn won't have LinearSVC. Good to have options.

File: tests/pipeline/test_mpca_trainer.py [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+@pytest.mark.parametrize("params", PARAMS)
+def test_mpca_trainer(classifier, params):
+    trainer = MPCATrainer(classifier=classifier, **params)
+    trainer.fit(x, y)
+    y_pred = trainer.predict(x)
+    testing.assert_equal(np.unique(y), np.unique(y_pred))
+    assert accuracy_score(y, y_pred) >= 0.8
+
+    if classifier == "linear_svc":
+        with pytest.raises(Exception):
```
Comment (haipinglu) [inline tests/pipeline/test_mpca_trainer.py:30]: Very glad to learn this from you. Thanks.
This may be another important documentation for developers to refer to (besides fixture): https://docs.pytest.org/en/stable/assert.html#assertions-about-expected-exceptions

**Review by haipinglu — APPROVED** [review]

**Comment by haipinglu** [comment]
@sz144 In-line docstrings seem fine but please **have documentation in `docs` updated _before merging_**. Thanks. 
This is actually the last checkbox in the description that you leave unchecked/ignored. You should complete and tick the applicable checkboxes when done. The checkbox is a checklist for you to review and check.

**Comment by sz144** [comment]
> 
> 
> @sz144 In-line docstrings seem fine but please **have documentation in `docs` updated _before merging_**. Otherwise, this new API won't appear in docs. Thanks.
> This is actually the last checkbox in the description that you leave unchecked/ignored. You should complete and tick the applicable checkboxes when done. The checkbox is a checklist for you to review and check.

Thanking you for pointing this out. The last checkbox is checked. Is there anything I need to do before merging?

**Comment by haipinglu** [comment]
> > @sz144 In-line docstrings seem fine but please **have documentation in `docs` updated _before merging_**. Otherwise, this new API won't appear in docs. Thanks.
> > This is actually the last checkbox in the description that you leave unchecked/ignored. You should complete and tick the applicable checkboxes when done. The checkbox is a checklist for you to review and check.
> 
> Thanking you for pointing this out. The last checkbox is checked. Is there anything I need to do before merging?

**Come on.** Have you done the _docs update_? I have make the text **have documentation in `docs` updated** bold. How can you tick something that you have not done yet? Could you check the docs yourself before asking me to check?

File: kale/pipeline/mpca_trainer.py [hunk 13]
```diff
@@ -0,0 +1,165 @@
...
+
+class MPCATrainer(BaseEstimator, ClassifierMixin):
+
+    def __init__(self, classifier='svc', classifier_params='auto', mpca_params=None,
+                 n_features=None, search_params=None):
+        """Trainer of pipeline: MPCA->Feature selection->Classifier
+
+        Args:
+            classifier (str, optional): Classifier for training. Options: support vector machine (svc) or
+                logistic regression (lr). Defaults to 'svc'.
```
Comment (haipinglu) [inline kale/pipeline/mpca_trainer.py:39]: Outdated docstrings. Now there are three options and you need to explain a bit so users know the difference from the docs rather than have to read the code.

File: kale/pipeline/mpca_trainer.py [hunk 14]
```diff
@@ -0,0 +1,165 @@
+# =============================================================================
+# Author: Shuo Zhou, szhou20@sheffield.ac.uk
+#         Haiping Lu, h.lu@sheffield.ac.uk or hplu@ieee.org
+# =============================================================================
+
+"""Implementation of MPCA->Feature Selection->Linear SVM/LogisticRegression Pipeline
```
Comment (haipinglu) [inline kale/pipeline/mpca_trainer.py:6]: Add references of papers using this pipeline: 
For Cardiac MRI: https://doi.org/10.1093/ehjci/jeaa001
For brain fMRI: https://doi.org/10.1007/978-3-319-24553-9_75
For gait videos (though KNN rather than SVM/LR was used): https://doi.org/10.1109/TNN.2007.901277

File: kale/pipeline/mpca_trainer.py [hunk 15]
```diff
@@ -0,0 +1,179 @@
...
+
+References:
+    [1] Swift, A. J., Lu, H., Uthoff, J., Garg, P., Cogliano, M., Taylor, J., ... & Kiely, D. G. (2020). A machine
+    learning cardiac magnetic resonance approach to extract disease features and automate pulmonary arterial
+    hypertension diagnosis. European Heart Journal-Cardiovascular Imaging.
+    [2] Song, X., Meng, L., Shi, Q., & Lu, H. (2015, October). Learning tensor-based features for whole-brain fMRI
+    classification. In International Conference on Medical Image Computing and Computer-Assisted Intervention
+    (pp. 613-620). Springer, Cham.
+    [3] Lu, H., Plataniotis, K. N., & Venetsanopoulos, A. N. (2008). MPCA: Multilinear principal component analysis of
+    tensor objects. IEEE transactions on Neural Networks, 19(1), 18-39.
```
Comment (haipinglu) [inline kale/pipeline/mpca_trainer.py:16]: "transactions" needs to be capitalized.
